### PR TITLE
ImproveArtemisDurationFromSecondsPipe to support more time formats

### LINKED ISSUE #3692: Exam overview: Scaling time for a "large" remaining time / duration time
### Is your feature request related to a problem?

In exams, the minute time should also be scaled up (to hours and days iff needed) and perhaps the date with time for end of exam should be displayed accordingly.

This is important for large remaining times (e.g. multiple days)

### Describe the solution you'd like

In exams, the minute time should also be scaled up (to hours and days iff needed) and perhaps the date with time for end of exam should be displayed accordingly.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JohannesStoehr**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I translated all newly inserted strings into English and German.

### Motivation and Context

Fix #3692

### Description

This PR changes the ArtemisDurationFromSecondsPipe to handle time spans that are multiple hours or even days long. The days and hours are left out, if they would only show 0.

This PR additionally fixes a type error that occurred in the refactored code.

### Steps for Testing

1. Log in to Artemis
2. Create an exam that has a long working time (>= 60 Minutes)
3. Participate as a student and check all occurrences of a duration for their correctness.
- The exam card in the exams tab of a course
- The exam cover sheet (If the student has additional working time, the amount of time added should be displayed in brackets)
- The timer in the top right of the exam and the hand in screen (The timer should only show the seconds when it is below ten minutes to not distract the students too much during the exam)
4. Check the localization of "day"/"Tag", if the exam spans multiple days

For the type error:
1. Go to the student exams and select one
2. View the submission of the exam
3. Check the console and make sure that no type error occurs.

### Test Coverage

artemis-duration-from-seconds.pipe.ts: Now tested 100%

### Screenshots

Exam timer:
Bildschirmfoto 2021-07-28 um 18 21 33
Bildschirmfoto 2021-07-28 um 18 22 02

Exam overview:
Bildschirmfoto 2021-07-29 um 12 53 14

Cover page (with additional working time displayed in brackets):
Bildschirmfoto 2021-07-29 um 12 52 47

### CODE REVIEW

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FuchsDominik) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:83]:
```suggestion
        if (this.locale === 'de') {
            return days > 1 ? ' Tage ' : ' Tag ';
        } else {
            return days > 1 ? ' days ' : ' day ';
        }
```
Small suggestion to shorten the code a bit

**Review by FuchsDominik — DISMISSED** [review]
Code looks good 👍

**Review by iozbeyli — COMMENTED** [review]
It still shows numbers like 7410 on the exam card in course overview.

**Review by MDK5 — DISMISSED** [review]
tested during testing session, two issues arose:
grafik
here the number is not formatted
grafik
it still says mm:ss even though its d hh:mm:ss

**Review by ndricimrr — COMMENTED** [review]
Tested on TS1. Noticed same issue as mentioned above on exam card

**Review by MDK5 — DISMISSED** [review]
Tested on ts1, the numbers on the overview and detail page don't match:
grafik
grafik

I am not sure why, but I think they should always show the same number

Its possible this is because the exam already started and one of the numbers refers to the remaining time, but then the label is misleading

**Review by willergeorg — DISMISSED** [review]
Tested on TS1. I was not able to reproduce the issue @MDK5 had. Format was appropriate and numbers were consistent.

**Review by NicolasUllmann — DISMISSED** [review]
Tested on TS1.  Everything worked for me as well.

**Review by FlorianGlombik — CHANGES_REQUESTED** [review]
1. I found an inconsistency: The time should be displayed in the same way, which means 1:50 pm in bought cases / 13:50 in bought cases - should differ depending on the selected language.
English version
  image
  German version
  image

2. The issue that @MDK5 addressed is reproducible, I would suggest changing the translation for the view in the Exam overview to `Remaining working time`
image
Also, note that the time format should be pm here as well.

3. Are the seconds really relevant?
image

File: src/main/webapp/app/overview/course-exams/course-exam-detail/course-exam-detail.component.ts
Comment (FlorianGlombik) [inline src/main/webapp/app/overview/course-exams/course-exam-detail/course-exam-detail.component.ts:28]: I think the comment is outdated 
```suggestion
     * calculate the duration in seconds between the start and end date of the exam
```

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FlorianGlombik) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:22]:
```suggestion
     * Convert seconds to a human-readable duration format "d day(s) hh:mm:ss".
```

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FlorianGlombik) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:89]:
```suggestion
        if (this.onLangChange !== undefined) {
```

File: src/main/webapp/i18n/de/exam.json
Comment (FlorianGlombik) [inline src/main/webapp/i18n/de/exam.json:40]: The `durationWithFormat` translation is now the same as the `exam.overview.duration`, I would suggest removing the first one and exchange it with the second translation that could be moved up one level.

File: src/main/webapp/i18n/en/exam.json
Comment (FlorianGlombik) [inline src/main/webapp/i18n/en/exam.json:40]: Same applies in the english translation https://github.com/ls1intum/Artemis/pull/3710/files#r667333310

File: src/main/webapp/app/shared/pipes/artemis-date.pipe.ts
Comment (FrankeLukas) [inline src/main/webapp/app/shared/pipes/artemis-date.pipe.ts:147]: Can you revert this change please? (Just using `!=` here checks for both undefined and null, so is preferred.)

File: src/main/webapp/app/exam/participate/information/exam-information.component.html
Comment (FrankeLukas) [inline src/main/webapp/app/exam/participate/information/exam-information.component.html:27]: We try to avoid method calls from the template (as they are potentially executed every render frame). These values don't seem to change once an exam is passed to the component, so can we calculate the values once and then reference fields here?

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FrankeLukas) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:89]:
```suggestion
        if (this.onLangChange != undefined) {
```

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FrankeLukas) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:83]: This part here looks rather odd. If `lang` and `locale` are `undefined`, we will first set `locale` to `en`, then compare `en` to `undefined`, which are different, and thus set the `locale` back to `undefined`?

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (FrankeLukas) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:76]: Is it possible to use the translation service here? (That would be better than hardcoding `de`.)

**Review by FrankeLukas — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exam/participate/information/exam-information.component.html
Comment (FrankeLukas) [reply src/main/webapp/app/exam/participate/information/exam-information.component.html:27]: Actually we can probably ignore this since the pipeline runs over the return value.

File: src/main/webapp/app/exam/participate/timer/exam-timer.component.ts
Comment (FrankeLukas) [inline src/main/webapp/app/exam/participate/timer/exam-timer.component.ts:7]: This import is unused now.

**Review by NicolasUllmann — APPROVED** [review]
Tested on TS1. re-approve.

**Review by FlorianGlombik — COMMENTED** [review]
I like that the seconds are not ticking down and not distracting the student when more than 10 minutes are left. On the other hand, the current solution with always displaying 00 seconds might be confusing and give students the feeling that something would not work.

I think just displaying the time in an `(d) h:m` format for more than 10 minutes left might be the best? Maybe an on hover could the included that reflects the current format?

A quick solution could also be to always display the seconds - in the end, some students that might be stressed out a bit more by the seconds ticking down might have less trouble than students constantly refreshing the page to update the timer? 

I am not sure if I am thinking too complicated here - some opinions on this would be beneficial, so feel free to add your comment.

**Review by FlorianGlombik — CHANGES_REQUESTED** [review]
If the current solution is chosen, I think at least an on hover explanation should be displayed that the seconds will not tick down as long as more than 10 minutes are left for the exam.

**Review by PatZintel — DISMISSED** [review]
tested on ts1 in testing session

**Review by jhainzlschmid — DISMISSED** [review]
Tested in the testing session on ts1

**Review by FlorianGlombik — DISMISSED** [review]
The time formats are now displayed in an intuitive and useful way (tested for an exam lasting more than a day, an exam lasting an hour, and an exam lasting only a few minutes)

**Review by canyelten — APPROVED** [review]
Tested on TS1, works as expected.

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (MDK5) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:45]: It seems a bit odd to use 1d 1h 1min 1s and not m for minute. I know this was like that before, so it should not block merging, it just seems odd.

**Review by MDK5 — DISMISSED** [review]
Tested on ts1, all durations looked good to me. Code changes also look good to me.

File: src/main/webapp/app/exam/participate/information/exam-information.component.ts
Comment (krusche) [inline src/main/webapp/app/exam/participate/information/exam-information.component.ts:27]: I think it would be better to move these into util files or in some kind of exam date service so that they can be reused in different views / components. Please also add some JavaDoc comments

**Review by ArtemTsurikov — APPROVED** [review]
Tested on TS1, looks good!

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:37]: 'min' is commonly used for minutes, where other time units are often just abbreviated by a single letter.
I guess 'm' clashes with the Meter...

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (lschlesinger) [inline src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:23]: I remember something such as "Please use [`Math.round`](https://github.com/ls1intum/Artemis/blob/fe5ab6c42bb64c9d6eabef8b58a131b37da3db80/src/main/webapp/app/shared/util/utils.ts#L60) from our `utils.ts` file" - I think this might apply here as well?

**Review by lschlesinger — APPROVED** [review]
I realized, that when I want to give a student extra working time on an exam that spans over multiple days, I am missing the right tool... I have to change the working time from, lets say 1d 0min 0sec to 1d 30min 0 sec if I want to grant 30 extra minutes. The interface for changing the individual working times (i.e. to include extra time) is meant for mm:ss input:
Bildschirmfoto 2021-07-29 um 14 25 33
I had to type in a high number of minutes to achieve what I wanted. But that can be tackled in another PR I guess.
The things that are described in the testing steps worked as expected. Also the type error was fixed when I tested on TS1.
The comments regarding the code are also not critical, if it has to be merged soon.

**Review by MDK5 — DISMISSED** [review]
Tested on ts1, worked as decibed. Code changes also look good to me

**Review by Olimbia23 — DISMISSED** [review]
Tested on TS1, works as should. @JohannesStoehr  👍 
Also I reviewed the code and left some comments.

File: src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/shared/pipes/artemis-duration-from-seconds.pipe.ts:23]: As long as I just want to round to the next integer, Math.round is having exactly the way as round() in utils.ts behaves.
(See first lines in that method)

File: src/main/webapp/app/exam/participate/exam-utils.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/exam/participate/exam-utils.ts:18]: But they can still be null, can't they?

**Review by MDK5 — APPROVED** [review]
Re-Approve, no functional changes

**Review by FlorianGlombik — APPROVED** [review]
Re-approving after reviewing the code changes since the last review, which have been method name refactorings, minor style issues, and comments that should not change the overall behavior.

**Pull request merged**
